File: frontend/lib/events.ts
// Shared types + client for the events news feed.
//
// Mirrors the backend Event schema (app/schemas/events.py). The browser calls
// same-origin /api/events, which proxies to the backend GET /events.

export type EventCategory =
  | "sports"
  | "festival"
  | "concert"
  | "construction"
  | "closure"
  | "weather";

export type EventKind = "demand_surge" | "supply_disruption";
export type Magnitude = "low" | "medium" | "high" | "severe";

export interface Venue {
  name: string;
  lat: number | null;
  lon: number | null;
}

export interface TransitImpact {
  magnitude: Magnitude;
  expected_attendance?: number | null;
  radius_km?: number | null;
  affected_lines: string[];
  affected_stations: string[];
  shuttle_replacement?: boolean | null;
}

export interface CityEvent {
  id: string;
  title: string;
  category: EventCategory;
  kind: EventKind;
  venue: Venue;
  start: string; // ISO datetime
  end: string;
  impact: TransitImpact;
  description: string;
  source: string;
}

export interface EventsResponse {
  as_of: string;
  window_days: number;
  count: number;
  events: CityEvent[];
}

/** Display metadata per category: short label + an emoji marker. */
export const CATEGORY_META: Record<EventCategory, { label: string; icon: string }> = {
  sports: { label: "Sports", icon: "🏟️" },
  festival: { label: "Festival", icon: "🎉" },
  concert: { label: "Concert", icon: "🎵" },
  construction: { label: "Construction", icon: "🚧" },
  closure: { label: "Closure", icon: "⛔" },
  weather: { label: "Weather", icon: "🌧️" },
};

/** Accent color per magnitude (for the severity dot/border). */
export const MAGNITUDE_COLOR: Record<Magnitude, string> = {
  low: "#34d399",
  medium: "#fbbf24",
  high: "#fb923c",
  severe: "#f87171",
};

/** Fetch upcoming events from the same-origin proxy. */
export async function loadEvents(): Promise<EventsResponse> {
  const res = await fetch("/api/events");
  if (!res.ok) throw new Error(`Failed to load events (HTTP ${res.status})`);
  return res.json();
}

File: backend/app/schemas/events.py
"""Schemas for upcoming city events that perturb transit.

An *event* is anything with a place and a time window that shocks the grid —
a stadium match (demand surge) or a line closure (supply disruption). The
`TransitImpact` block carries the *transparent assumptions* a shock model
consumes; they are deliberately coarse and human-readable, never hidden
predictions — a downstream shock model turns them into grid deltas.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class EventCategory(str, Enum):
    sports = "sports"
    festival = "festival"
    concert = "concert"
    construction = "construction"
    closure = "closure"
    weather = "weather"


class EventKind(str, Enum):
    """Which side of the system the event shocks."""

    demand_surge = "demand_surge"          # more riders at a place/time
    supply_disruption = "supply_disruption"  # less network capacity


class Magnitude(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"
    severe = "severe"


class Venue(BaseModel):
    """Where the event is centred. Coords are WGS84; absent for diffuse events."""

    name: str
    lat: Optional[float] = None
    lon: Optional[float] = None


class TransitImpact(BaseModel):
    """Coarse, transparent shock parameters — the inputs to a shock model.

    These are assumptions, shown to the user, not forecasts. A downstream
    `model_event_shock` tool turns them into grid deltas.
    """

    magnitude: Magnitude = Magnitude.medium
    # demand-surge params
    expected_attendance: Optional[int] = Field(
        None, description="Approximate peak crowd, for demand surges."
    )
    radius_km: Optional[float] = Field(
        None, description="How far the surge spreads from the venue."
    )
    # supply-disruption params
    affected_lines: list[str] = Field(default_factory=list)
    affected_stations: list[str] = Field(default_factory=list)
    shuttle_replacement: Optional[bool] = Field(
        None, description="Whether shuttle buses replace the closed segment."
    )


class Event(BaseModel):
    """A time-bounded event with an inferred transit impact."""

    id: str
    title: str
    category: EventCategory
    kind: EventKind
    venue: Venue
    start: datetime
    end: datetime
    impact: TransitImpact = Field(default_factory=TransitImpact)
    description: str = ""
    source: str = Field("mock", description="Provenance — 'mock' until real feeds wired in.")
